frameworklauncher/src/main/java/com/microsoft/frameworklauncher/applicationmaster/SelectionManager.java

```diff
@@ -46,185 +45,217 @@
+        if (totalResource.getGpuNumber() > 0) {
+          LOGGER.logDebug("skip nodes with Gpu resource for 0-gpu job: Node [%s], Node total resource: [%s]",
+              node.getHost(), totalResource);
+          filteredNodes.remove(i - 1);
+        }
       }
+    }
+  }
 
-      // Test Optimized Resource
-      ResourceDescriptor optimizedRequestResource = YamlUtils.deepCopy(requestResource, ResourceDescriptor.class);
-      if (requestResource.getGpuAttribute() == 0) {
-        // If GpuAttribute is not explicitly specified, we select an optimal GpuAttribute according to 
-        // the current status of the node instead of let RM to select a random GpuAttribute.
-        optimizedRequestResource.setGpuAttribute(selectCandidateGpuAttribute(node, requestResource.getGpuNumber()));
-        if (!ResourceDescriptor.fitsIn(optimizedRequestResource, availableResource)) {
-          LOGGER.logDebug(rejectedLogPrefix +
-                  "Resource does not fit in: Optimized Request Resource: [%s], Available Resource: [%s]",
-              optimizedRequestResource, availableResource);
-          continue;
+  private void filterNodesByResource(ResourceDescriptor requestResource, Boolean skipLocalTriedResource) {
+    if (requestResource != null) {
+      for (int i = filteredNodes.size(); i > 0; i--) {
+        Node node = allNodes.get(filteredNodes.get(i - 1));
+        ResourceDescriptor availableResource = node.getAvailableResource();
+        if (skipLocalTriedResource && localTriedResource.containsKey(node.getHost())) {
+          LOGGER.logDebug("Skip local tried resources: [%s] on Node : [%s]", localTriedResource.get(node.getHost()), node.getHost());
+          ResourceDescriptor.subtractFrom(availableResource, localTriedResource.get(node.getHost()));
+        }
+        if (!ResourceDescriptor.fitsIn(requestResource, availableResource)) {
+          LOGGER.logDebug("Resource does not fit in: Node: [%s] Request Resource: [%s], Available Resource: [%s]",
+              node.getHost(), requestResource, availableResource);
+          filteredNodes.remove(i - 1);
         }
       }
+    }
+  }
+
+  private void filterNodesByRackSelectionPolicy(ResourceDescriptor requestResource, int pendingTaskNumber) {
+    //TODO: Node GPU policy filter the nodes;
+  }
 
-      // Found a selectionResult passed all the Tests above
-      selectionResult = new SelectionResult();
-      selectionResult.setNodeHost(nodeHost);
-      selectionResult.setGpuAttribute(optimizedRequestResource.getGpuAttribute());
-      break;
+  private SelectionResult SelectNodes(ResourceDescriptor requestResource, int pendingTaskNumber) {
+    //TODO: apply other node selection policy in  the futher;
+    return selectNodesByJobPacking(requestResource, pendingTaskNumber);
+  }
+
+  //Default Node Selection strategy.
+  private SelectionResult selectNodesByJobPacking(ResourceDescriptor requestResource, int pendingTaskNumber) {
```

Comment:
>selectNodesByJobPacking [](start = 26, length = 23)

You should also need to selectNodes which can nearly maximize the overlaped ports.

Reply:
good point , not this PR

---
In reply to: [174757105](https://github.com/Microsoft/pai/pull/318#discussion_r174757105) [](ancestors = 174757105)